R/source_code.R

```diff
@@ -0,0 +1,114 @@
+    mod_out_r <- reactive({
+      if (!is.null(module_out) && is.reactive(module_out)) {
+        tryCatch(module_out(), error = function(e) e)
+      }
+    })
+
+    code_out <- reactive({
+      teal_data_handled <- mod_out_r()
+      if (inherits(teal_data_handled, "qenv")) {
+        teal.code::get_code(teal_data_handled)
+      }
+    })
+
+    # Global option that doesn't show button
+    is_button_showed_r <- shiny::reactive(getOption("teal.show_src", TRUE))
+
+    reason_r <- reactive({
+      if (is.null(mod_out_r())) {
+        "No source code is available from this module."
+      } else if (inherits(mod_out_r(), "error")) {
+        "The module returned an error, check it for errors."
+      } else if (is.null(code_out())) {
+        "The module does not support source code functionality"
+      } else if (isFALSE(attr(mod_out_r(), "teal.enable_src"))) {
+        "The show source code functionality is disabled for this module."
+      }
+    })
+
+    output$source_code_container <- renderUI({
+      if (is_button_showed_r()) {
+        bslib::tooltip(
+          id = session$ns("source_code_tooltip"),
+          trigger = shiny::tags$div(
+            id = session$ns("source_code_wrapper"),
+            class = "cursor-helper",
+            ui_source_button(session$ns(NULL))
+          ),
+          ""
+        )
+      }
+    })
+
+    teal.widgets::verbatim_popup_srv(
+      id = "source_code", verbatim_content = code_out, title = "Show R Code for Response"
+    )
+
+    observeEvent(code_out(), ignoreNULL = FALSE, {
+      reason <- trimws(reason_r() %||% "")
+      if (is.null(reason) || identical(reason, "")) {
+        session$sendCustomMessage("disable-tooltip", session$ns("source_code_wrapper"))
+      } else {
+        session$sendCustomMessage("enable-tooltip", session$ns("source_code_wrapper"))
+        bslib::update_tooltip(id = "source_code_tooltip", reason)
+      }
+
+      shinyjs::toggleState(
+        "source_code_wrapper",
+        condition = !is.null(code_out()) &&
+          inherits(mod_out_r(), "qenv") &&
+          !isFALSE(attr(mod_out_r(), "teal.enable_src")) # Only forcibly disable when value is explicitly FALSE
+      )
+    })
+  })
+}
+
+#' Disable the "Show R Code" global button in the UI
+#'
+#' Convenience function that disables the user's ability to see the code of the module.
+#' @param x (`teal_module`) a `teal_module` object.
+#' @return modified data object that indicates that it should not show the "Show R Code"
+#' button in the UI.
+#' @examples
+#' app <- init(
+#'   data = within(teal_data(), iris <- iris),
+#'   modules = modules(
+#'     example_module(label = "example teal module") |> disable_src()
+#'   )
+#' )
+#' if (interactive()) {
+#'   shinyApp(app$ui, app$server)
+#' }
+#' @export
+disable_src <- function(x) {
+  checkmate::assert_multi_class(x, c("teal_module", "teal_modules"))
+  after(x, server = function(data) {
+    attr(data, "teal.enable_src") <- FALSE
```

Comment:
Discussion for @insightsengineering/nest-core-dev (from coredev corner)

Using attributes set on the S4 object instead of the slots. The attribute names uses a namespace `prefix` to indicate the context.

In the future we may use `shared.xxxx` to pass options to other modules (think `mdr` composite modules). This makes it easier to maintain over time.